### DISCUSSION TITLE
AzureMonitor: strongly-typed JS clients for Azure APIs

### DISCUSSION THREAD

**Discussion by kostrse** [grafana/grafana#41983] [Data Sources]
All API requests to Azure (Azure Management API etc.) in the Azure Monitor datasource frontend code should be refactored into a strongly-typed client (wrapper) with accurate models:
* Create an Azure API client class which provides common API methods like `getSubscriptions`, `getResourceGroups` etc.
* Use the client from all placed which currently craft API requests directly (datasources, controls etc.).
* Have accurate typings of returned data structures.
* Potentially, reuse the strongly-typed client across different Azure datasources via a shared library.

Also, investigate if it's possible to use the official [Azure SDK for JavaScript](https://github.com/Azure/azure-sdk-for-js) library with customized transport which would forward all HTTP requests to the datasource proxy using HTTP client provided by Grafana. This would free us from maintaining the up-to-date data models of Azure APIs.

**Comment by joshhunt** [comment]
When someone works on this I would start with investigating use of Azure SDK to talk to the Azure API through Grafana, exactly as @kostrse mentioned,

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
